dspace-uclouvain/src/main/java/org/dspace/uclouvain/submissionMetadataGenerators/generators/DegreeMetadataGenerator.java

```diff
@@ -34,15 +51,49 @@ public String getGeneratorName() {
     public void process(Context ctx, Item item) throws GeneratorProcessException {
         try {
             ItemService currentItemService = item.getItemService();
-            List<MetadataValue> currentDegreeCodes = currentItemService.getMetadata(item, this.degreeCodeField.getSchema(), this.degreeCodeField.getElement(), this.degreeCodeField.getQualifier(), null);
-            for (MetadataValue mv: currentDegreeCodes) {
-                List<String> newValues = this.generateNewDegreeCode(mv);
+            // 1) Clear previous metadata Values
+            this.clearPreviousMetadata(ctx, item, currentItemService);
+
+            // 2) Retrieve a list of distinct degree codes and labels splitted by '-'
+            List<MetadataValue> currentDegreeCodes = currentItemService.getMetadata(
+                item,
+                this.degreeCodeField.getSchema(),
+                this.degreeCodeField.getElement(),
+                this.degreeCodeField.getQualifier(),
+                null
+            );
+            List<List<String>> codeAndLabels = this.generateCodeAndLabels(currentDegreeCodes);
+            for (Integer elementIndex = 0; elementIndex < codeAndLabels.size(); elementIndex++) {
```

Comment:
Add a comment here to explain what you do. In 6 months (weeks), we no longer understand why this loop.